### ISSUE TITLE
fails for symmetrized HKUST-1 structure

### ISSUE THREAD

**Issue by kjappelbaum** [coudertlab/CrystalNets.jl#5]
Hi, i didn't trace back yet where exactly it fails (somewhere in the net extraction) but for the symmetrized (i.e., non P1) version of HKUST-1 it fails with the error that the net is unstable (even though i would still expect it to return tbo). 

Attached the cif. I'll keep digging a bit to understand what is the reason for it. 

[hkust_sym.cif.txt](https://github.com/coudertlab/CrystalNets.jl/files/5963768/hkust_sym.cif.txt)

**Comment by Liozou** [comment]
Hi,
Thanks for the bug report! There are multiple improvements and fixes in the branch `multipledimensions`, which I plan on merging on master as soon as I can... But in the meantime you may want to try it and see if the bug was already fixed there.
Otherwise I'll look into it and let you know.

**Comment by Liozou** [comment]
Alright, I inspected your input and there is actually a difference: the two Cu atoms from the paddle-wheel are not bound together in your input, whereas it is in the CIF files from the `test` folder. This is not related to the symmetry consideration. However, it makes a difference when you look at the automatically determined SBUs:
- In the CIF file from `test`, since the two Cu atoms have a bond, both are assumed to belong to the same (inorganic) SBU, which is indeed correct.
- In your input file however, the inorganic SBUs are split in two, each half containing only one Cu atom. Since these twins vertices share the same 4 neighbors (which are the organic SBUs they are bound to), the graph becomes unstable.

At the moment, I don't see how to fix this general issue in the package, since it amounts to correctly guessing what atoms should be bound together and I don't have a better guess than Chemfiles. At least the warning is present (on the `multipledimensions` branch):
```
┌ Warning: Guessing bonds through Chemfiles. This may take a while for big structures and may be inexact.
└ @ CrystalNets ~/.julia/dev/CrystalNets/src/input.jl:296
[ Info: To avoid guessing bonds, use a file format that contains the bonds.
```
And indeed, the solution for the user is to provide an input that contains the bonds. But that is a hassle in many cases.

I can see a fix for this particular case, which might turn out to be more relevant than just this. It consists in using the following heuristic when determining the topology of a MOF: if an atom from an organic SBU is bound to more than 1 inorganic SBUs, then all these inorganic SBUs are actually one and the same. Given the structure of a MOF, I think this is a valid heuristic in most cases, but let me know what you think of it if you have an idea! I will try to implement it and see if it makes sense otherwise.

**Comment by kjappelbaum** [comment]
Hi, sorry for coming back to it delayed and thanks for your efforts! 

>if an atom from an organic SBU is bound to more than 1 inorganic SBUs, then all these inorganic SBUs are actually one and the same.

If i get correctly what you're trying to do it will be ok in >99% of the cases but might fail for cases in which on has mixed-metal nodes in with stochastic arrangements of the metals (the connectivity of the graph would still be same, only the node labels change ...) 

And detecting the bond network is indeed a really difficult task. For MOFs, the CrystalNN algorithm often does a decent job (https://pubs.acs.org/doi/10.1021/acs.inorgchem.0c02996).

**Comment by Liozou** [comment]
No problem! I implemented what I proposed in https://github.com/coudertlab/CrystalNets.jl/commit/32a951892f450152fd45099fe367b0f8b79851bb so you should have access to it if you work on branch [multipledimensions](https://github.com/coudertlab/CrystalNets.jl/tree/multipledimensions). I also fixed the issue for which you submitted PR #4 on that branch, I'll close the PR (and this issue) once the branch is merged back to master.

> cases in which on has mixed-metal nodes in with stochastic arrangements of the metals

I'm not sure to know whether those case are correctly handled anyway, since I don't have a well-defined heuristic when multiple atoms collide. But if the connectivity of the graph is the same in the end, then its detected topology should be also be the same. If you have an input file describing the issue you foresee, don't hesitate to send it and I'll have a look.

Thanks for the recent link on bond detection, it looks promising!

**Comment by Liozou** [comment]
There has been a number of changes in between, but now the behavior should be correct: I get
```julia
julia> determine_topology("hkust_sym.cif", structure=StructureType.MOF)
AllNodes, SingleNodes: tbo
```
The specific behavior of paddle-wheels is toggled by the `detect_paddlewheel` boolean keyword in the [options](http://molsim.info/CrystalNets.jl/dev/lib/public/#CrystalNets.Options).
